### PR TITLE
Prometheus: Check both error and http status for source of error 

### PR DESCRIPTION
**Pull request by bohandley**
An improved version of this [reverted PR](https://github.com/grafana/grafana/pull/95999) [using this doc](https://docs.google.com/document/d/1FWbHFg6aEYgCo9uSvDaTCPcb9xUub-3rzXpBGjudWxU/edit?tab=t.0) written by @ivanahuckova 

In this, we check the source of the error as well as the status code before labeling as either from downstream or from our code for the plugin.

We use the helper functions below:

`backend.IsDownstreamHTTPError(err)`: check the error to see if the source is downstream or other
`backend.ErrorSourceFromHTTPStatus(res.StatusCode)`: check the status code for error source, down stream or other

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Comment by itsmylife** [comment]
@bohandley It looks quite nice! Nice job!
I see that we repeat some code here and there. Can we do the same without duplicating the code?

**Review by itsmylife — APPROVED** [review]

File: pkg/promlib/querydata/request.go [hunk 1]
```diff
@@ -234,9 +234,14 @@ func (s *QueryData) fetch(traceCtx context.Context, client *client.Client, q *mo
...
 	if err != nil {
+		return addErrorSourceToDataResponse(err)
+	}
+
+	if res.StatusCode/100 != 2 {
+		// for differentiating the source of http errors by status code
 		return backend.DataResponse{
-			Error:  err,
-			Status: backend.StatusBadGateway,
+			Error:       err,
```
Comment (ivanahuckova) [inline pkg/promlib/querydata/request.go:243]: I think that in this case error might be `nil` and you should check the response in what form are we getting the error message. This is what we use in Loki https://github.com/grafana/grafana/blob/e86283e8b173b328c4710720b8f807f359f57a77/pkg/tsdb/loki/api.go#L145  and https://github.com/grafana/grafana/blob/e86283e8b173b328c4710720b8f807f359f57a77/pkg/tsdb/loki/api.go#L126

**Review by ivanahuckova — COMMENTED** [review]
Left a comment bellow related to possible nil error message, but the error source work is good and that part LGTM!

**Comment by itsmylife** [comment]
Just want to understand something, after @ivanahuckova's comment, the `Error` was removed but the `ErrorSource` is stayed. If we don't have error in the response why do we have the `ErrorSource`? 
Prometheus might return 3 different status codes other then 200. 400, 422 and 503. https://prometheus.io/docs/prometheus/latest/querying/api/#format-overview
If we have them, the `err` from query call would be `nil` but response would have `errorType` and `error`. And I think we should extract that error message and return it to the client.

**Comment by bohandley** [comment]
> Just want to understand something, after @ivanahuckova's comment, the `Error` was removed but the `ErrorSource` is stayed. If we don't have error in the response why do we have the `ErrorSource`? Prometheus might return 3 different status codes other then 200. 400, 422 and 503. https://prometheus.io/docs/prometheus/latest/querying/api/#format-overview If we have them, the `err` from query call would be `nil` but response would have `errorType` and `error`. And I think we should extract that error message and return it to the client.

@itsmylife update for this here!
https://github.com/grafana/grafana/pull/96347

**Pull request merged**